crates/cadenza-eval/src/diagnostic.rs

```diff
@@ -107,6 +110,81 @@ pub enum DiagnosticKind {
     InternalError(String),
 }
 
+/// Represents what type(s) were expected in a type error.
+///
+/// This can be a single type, multiple types, or a custom description
+/// for more complex type expectations (like "number" meaning integer or float).
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum TypeExpectation {
+    /// A single specific type was expected.
+    Single(Type),
+    /// One of several types was expected.
+    OneOf(Vec<Type>),
+    /// A custom description of expected types.
+    Description(String),
```

Comment:
I don't want custom strings. We should be able to describe any type expectation with the type system. If we can't then it's not extensive enough and we need to add to it.

Actually same goes for the OneOf. My suggestion was to add a union type that can express a list of accepted types.